src/mmdata.c

```diff
@@ -2413,16 +2464,48 @@ long g_pntrStartTempAllocStack = 0;   /* Where to start freeing temporary alloca
                                     when pntrLet() is called (normally 0, except in
                                     special nested vstring functions) */
 /*!
- * \var pntrTempAllocStack
- * \brief a \ref stack "stack" of \a temp_pntrString.
+ * \var pntrString *pntrTempAllocStack[]
+ * \brief a \ref stack "stack" of \ref temp_pntrString.
  *
- * Holds pointers to temporarily allocated data of type \a pntrString.  Such a
- * \ref stack "stack" contains strictly __local__ data of a function, not
- * accessed from outer levels.
+ * Holds pointers to temporarily allocated data of type \ref temp_pntrString.  Such
+ * a \ref stack "stack" is primarily designed to operate like one for temporary
+ * allocated ad hoc operands, as described in \ref stack.  The stack top index
+ * is \ref g_pntrTempAllocStackTop, always refering to the next push position.
+ * The \ref g_pntrStartTempAllocStack supports nested operations by indicating
+ * where the operands for the upcoming operation start from.
+ * \attention A \ref pntrString consists of an array of pointers.  These
+ *   pointers may themself refer data that needs a clean up, when the last
+ *   reference  to it disappears (such as deallocating memory for example).
+ *   There is no automatic procedure handling such cases when pointers are
+ *   popped off the stack to be freed.
+ * \bug The element type should be temp_pntrString, because a NULL_PNTRSTRING
+ *   must not be pushed on the stack.
  */
 pntrString *pntrTempAllocStack[M_MAX_ALLOC_STACK];
 
-
+/*!
+ * \fn temp_pntrString *pntrTempAlloc(long size)
+ * \par size > 0
+ * allocates a \ref block capable of holding \p size \ref pntrString entries
+ * and pushes it onto the \ref pntrTempAllocStack.
+ * \par size == 0
+ * pops off all entries from index \ref g_pntrStartTempAllocStack on from
+ * \ref pntrTempAllocStack and adds them to the \ref memFreePool.
+ * \param[in] size count of \ref pntrString entries.  This value must include
+ *   a terminal NULL pointer if needed.
+ * \return a pointer to the allocated \ref block, or NULL if deallocation
+ *   requested
+ * \pre
+ *   \p size ==0: all entries in from \ref pntrTempAllocStack from
+ *   \ref g_pntrStartTempAllocStack do not contain relevant data any more.
+ * \post
+ *   - \p size > 0: memory for \p size entries is reserved in the \ref block
```

Comment:
Again a best practice issue...  But you are right, the results for \p are not that convincing.  It is just standard.  Added this to the todos.

Reply:
There is a general principle ruling documentation systems (and to some lesser extent modern web development, too, see CSS guide lines):
The contents is strictly separated from its formatting.  The person writing the contents must not be concerned with how it is presented on different media like HTML, PDF or Latex. This looks perhaps a bit weird to Microsoft Word addicts, who are used to paint individual sentences much to their liking - some in green, others in red and so on.  Professionals have a different attitude: If a word is underlined, there is a reason for it, and only that reason is given by a semantic mark.  The underlining is finally added by the documentation system. That way all of the papers appear consistently styled, and the reader intuitively knows how to interpret the formatting after a few pages..
There is even more than conformity to it:  Should you ever upgrade to a next-level-fancy-eyecandy system, chances are, the input of doxygen can be automatically interpreted, if it depends on semantic mark-up only.
So how should something like "size > 0" be handled? This is a bit of a border case. On one hand is size a parameter and should receive the \p prefix.  But it is used in a formula context, which is tagged differently.  I am sure people with more experience can tell: You are in the long run better off using ... .  I am not.  I feel a bit reluctant to change the size > 0 (and only that!) for a bit of change in appearance.

Reply:
I have a simple rule: code or "math" text goes in `code` spans. This is semantic markup in the sense you are referring to: Using code in text like > 0 will cause it to be interpreted by renderers, screen readers, etc. as English text, but it isn't. (If you wanted English text you would write "greater than zero".) That goes even more so for size == 0 since `==` is a clear C-ism. The only time I make an exception to that rule is for linked code spans, but you can still work around it to some extent such as [`size`](https://github.com/metamath/metamath-exe/pull/75)` > 0`.

Unresolving the conversation because there are a ton of resolved conversations on this page that are still accumulating responses, and github makes it hard to find them when you do this.

Reply:
In src/mmdata.c, line 2515: 
> I have a simple rule: code or "math" text goes in code spans.

I have another simple rule: Prefer semantic tags over formatting

> The only time I make an exception to that rule is for linked code spans,..

This occurs more frequently than perhaps expected: Here is an example `\invariant \ref g_startTempAllocStack <= \ref g_tempAllocStackTop` in mmvstr.h, others could be "length + 1", and the "size == 0".  Whether a C-ism is acceptable in C source documentation, is again a matter of taste.

My problem is: This pull request is held up by, as it seems, formatting issues only.  There is no contention about contents like "You forgot this precondition" or "even for a draft this is too misleading".

That's why I do not unresolve this issue.  I want to document the source, and idealistically I need not care in the least about formatting, because the documentation system handles it transparently for me.

Ok reality isn't that forthcoming.  That's what I introduced the doc-todo page for.  Here suggestions can be gathered for regular review and rewrite in a single effort.

Reply:
The PR is held up by my not having time to review. If you don't have strong feelings about this, I am happy to merge it and make changes later. For your examples, it is possible to write those examples using words, i.e. "invariant: \ref g_startTempAllocStack is at most \ref g_tempAllocStackTop", or to use code and follow up with words: "invariant: `g_startTempAllocStack <= g_tempAllocStackTop`, where \ref g_startTempAllocStack and \ref g_tempAllocStackTop are global variables". Or you could just use <= in text or split the spans or something else. There are lots of options and I would take it on a case by case basis.

My worry is that you are taking a strong stance on this and would be opposed to changes to it later. There seems to be a mixture of things you haven't had time or the wherewithal to optimize presentation-wise in here, along with things written very deliberately in accordance with your philosophy of code documentation, and so I'm not sure whether to push back now or accept early and submit a change later addressing some particular thing like changing all the code spans.

Reply:
"My worry is that you are taking a strong stance on this.."  I am not opposed to changes of others.  This is not my first contribution to open source projects, and I can provide examples where my original text/code was rewritten by others both here and in other projects.  During the review of PR https://github.com/metamath/metamath-exe/pull/9 I was more than once completely at a loss to decide (for me) whether something was in a correct/acceptable shape or not.  I felt I should address this shortcoming, and started documenting the code (but deliberately not rewriting it!).  Documentation imo has too sides to it: 1. get the contents right; 2. present it in a way understandable to a majority.  Both aspects are important in the end, but whereas facts can be verified right from the start, best practices are subject to taste and maturing.  If I wait for the second aspect to be clarified in all detail in advance, I need not go on here for, say, a couple of months.  I rely to some extent on the documentation system Doxygen to provide a good and established starting point, even to a beginner with constraint time resources like me.

The documentation style with its pre-/postconditions and invariants focus on program state.  If done thoroughly enough, one can sort of prove the correctness of a program.  I thought this is appealing to this particular audience here.  Of course, one can advocate something different.  Well, move on then.